File: shared/db/schema/habits.ts
import { pgTable, uuid, varchar, timestamp, integer, real } from "drizzle-orm/pg-core";

/**
 * HABIT_DEFINITIONS Table
 * Defines the recurring habits the user wants to track.
 */
export const habitDefinitions = pgTable("habit_definitions", {
  id: uuid("id").primaryKey().defaultRandom(),
  title: varchar("title", { length: 255 }).notNull(),
  frequency: varchar("frequency", { length: 50 }).notNull(), // e.g., 'daily', 'weekly'
  unit: varchar("unit", { length: 20 }).default("rep").notNull(),
  two_min_threshold: integer("two_min_threshold").notNull().default(0),
  normal_threshold: integer("normal_threshold").notNull().default(0),
  hard_threshold: integer("hard_threshold").notNull().default(0),
  impossible_threshold: integer("impossible_threshold").notNull().default(0),
  points: integer("points").notNull().default(0),
  level: integer("level").notNull().default(0),
  goalId: uuid("goal_id"), // Optional reference to a Goals module (decoupled)
  createdAt: timestamp("created_at", { withTimezone: true }).defaultNow().notNull(),
  updatedAt: timestamp("updated_at", { withTimezone: true }).defaultNow().notNull(),
});

/**
 * HABIT_LOGS Table
 * Records instances of habit completion or status updates.
 */
export const habitLogs = pgTable("habit_logs", {
  id: uuid("id").primaryKey().defaultRandom(),
  habitId: uuid("habit_id")
    .references(() => habitDefinitions.id, { onDelete: "cascade" })
    .notNull(),
  value: real("value").notNull().default(0), // reps or minutes
  timestamp: timestamp("timestamp", { withTimezone: true }).defaultNow().notNull(),
});

File: backend/app/models/habits.py
from sqlalchemy import Column, String, Float, DateTime, ForeignKey
from sqlalchemy.orm import relationship
from .base import Base
from datetime import datetime, timezone
import uuid

class Habit(Base):
    __tablename__ = "habits"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    name = Column(String, nullable=False)
    unit = Column(String, default="mins")
    frequency = Column(String, default="daily")
    habit_type = Column(String, default="numeric") # numeric, binary
    two_min_threshold = Column(Float)
    normal_threshold = Column(Float)
    hard_threshold = Column(Float)
    impossible_threshold = Column(Float)
    
    logs = relationship("HabitLog", back_populates="habit", cascade="all, delete-orphan")

class HabitLog(Base):
    __tablename__ = "habit_logs"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    habit_id = Column(String, ForeignKey("habits.id"))
    value = Column(Float, nullable=False)
    timestamp = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))

    habit = relationship("Habit", back_populates="logs")
